src/Features/Core/Portable/ChangeSignature/ChangeSignatureCodeActionOperation.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Notification;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.ChangeSignature
+{
+    /// <summary>
+    /// Defines the <see cref="CodeActionOperation"/> for the <see cref="ChangeSignatureCodeAction"/>
+    /// This is used instead of <see cref="ApplyChangesOperation"/> as we need to show a confirmation
+    /// dialog to the user before applying the change.
+    /// </summary>
+    internal class ChangeSignatureCodeActionOperation : CodeActionOperation
+    {
+        public Solution ChangedSolution { get; }
+
+        public string? ConfirmationMessage { get; }
+
+        public ChangeSignatureCodeActionOperation(Solution changedSolution, string? confirmationMessage)
+        {
+            ChangedSolution = changedSolution ?? throw new ArgumentNullException(nameof(changedSolution));
+            ConfirmationMessage = confirmationMessage;
+        }
+
+        internal override bool ApplyDuringTests => true;
+
+        public override void Apply(Workspace workspace, CancellationToken cancellationToken)
+            => this.TryApply(workspace, new ProgressTracker(), cancellationToken);
+
+        /// <summary>
+        /// Show the confirmation message, if available, before attempting to apply the changes.
+        /// </summary>
+        internal override bool TryApply(
+            Workspace workspace, IProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            if (ConfirmationMessage != null)
+            {
+                var notificationService = workspace.Services.GetRequiredService<INotificationService>();
+                if (!notificationService.ConfirmMessageBox(ConfirmationMessage, severity: NotificationSeverity.Warning))
+                {
+                    return false;
```

Comment:
what does returning false do here higher up the chain?

Reply:
So I dug around a bit, and not much.  As best I can tell it's used only in integration tests to confirm if the code action application succeeded or not - http://sourceroslyn.io/#Microsoft.VisualStudio.IntegrationTest.Utilities/InProcess/TextViewWindow_InProc.cs,423

This behavior seems to match the other operations (e.g. InstallPackageAndAddImportOperation, InstallPackageDirectlyCodeActionOoperation)